theano/tensor/opt.py

```diff
@@ -3360,6 +3360,23 @@ def local_incsubtensor_of_zeros(node):
             return
 
 
+@register_canonicalize
+@register_stabilize
+@gof.local_optimizer([IncSubtensor])
+def incsubtensor_of_zeros_to_setsubtensor(node):
+    """
+    IncSubtensor(zeros, x, ...) -> SetSubtensor(zeros, x, ...)
+
+    """
+    if (isinstance(node.op, (IncSubtensor)) and not node.op.set_instead_of_inc):
+        x = node.inputs[0]
+        y = node.inputs[1]
+
+        if isinstance(x, T.zeros):
+            node.op.set_instead_of_inc = True
+        return node
```

Comment:
You should return a Theanno variable, not a new Apply node.

It should be something like:

    return IncSubtensor(node.op.idx_list, set_instead_of_inc=True, node.op.idx_list)(*node.inputs)